pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -895,9 +899,143 @@ func (t *Test) plan() (plan *TestPlan, retErr error) {
 	if retries > 0 {
 		t.logger.Printf("WARNING: generated a smaller (%d) test plan after %d retries", plan.length, retries)
 	}
+	// Sanity check.
+	if len(plan.upgrades) == 0 {
+		return nil, errors.Newf("internal error: plan has no upgrades\n%s", plan.prettyPrintInternal(true))
+	}
+	// Finally, check that user-specified hooks occur in the plan.
+	err := t.assertExpectedUserHooks(plan)
```

Comment:
New assertion to ensure we don't leave any hooks "hanging", i.e., each one must be used by the plan.